### PR TITLE
Added production environment

### PR DESCRIPTION
**Pull request by temitope777**
WHY: The prod environment is the most important and public facing
HOW: By creating Makefile entry and adding other tf resource files

## Context
Added the prod environment to ITTMS

## Changes proposed in this pull request
Added prod environment by making makefile entry and adding tf resources

## Guidance to review

make development terraform-kubernetes-apply 
## Link to Trello card
https://trello.com/c/3jvfiAei/1667-ittms-create-production-environment

## Things to check
ensure links work
ensure dfe sign in works
ensure monitoring works
ensure pipeline works

### CODE REVIEW

File: global_config/production.sh [hunk 1]
```diff
@@ -0,0 +1,5 @@
+CONFIG=prd
```
Comment (RMcVelia) [inline global_config/production.sh:1]: this should be "production"

File: terraform/application/application.tf [hunk 2]
```diff
@@ -56,6 +56,6 @@ module "worker_application" {
   cluster_configuration_map  = module.cluster_data.configuration_map
   kubernetes_config_map_name = module.application_configuration.kubernetes_config_map_name
   kubernetes_secret_name     = module.application_configuration.kubernetes_secret_name
-
+  replicas                   = var.sidekiq_replicas
```
Comment (RMcVelia) [inline terraform/application/application.tf:59]: Sidekiq isn't used, this should be var.worker_replicas

File: terraform/application/config/prd.tfvars.json [hunk 3]
```diff
@@ -0,0 +1,17 @@
+{
+  "cluster": "production",
+  "namespace": "bat-production",
+  "infra_key_vault_name": "s189p01-ittms-prd-inf-kv",
+  "statuscake_password_name": "SC-PASSWORD",
+  "key_vault_resource_group":  "s189p01-ittms-prd-rg",
+  "enable_monitoring" : true,
+  "app_replicas" : 2,
+  "sidekiq_replicas" : 2,
```
Comment (RMcVelia) [inline terraform/application/config/prd.tfvars.json:9]: worker_replicas, and confirm it's ok to have multiple replicas?

File: terraform/application/config/prd_Terrafile [hunk 4]
```diff
@@ -0,0 +1,5 @@
+aks:
+    source:  "https://github.com/DFE-Digital/terraform-modules"
+    version: "testing"
```
Comment (RMcVelia) [inline terraform/application/config/prd_Terrafile:3]: stable

File: terraform/application/application.tf (hunk 2 above)
Comment (temitope777) [reply terraform/application/application.tf:59]: changed to worker_replicas

File: terraform/application/config/prd_Terrafile (hunk 4 above)
Comment (temitope777) [reply terraform/application/config/prd_Terrafile:3]: changed to stable

File: terraform/application/config/prd.tfvars.json (hunk 3 above)
Comment (temitope777) [reply terraform/application/config/prd.tfvars.json:9]: changed to worker_replicas

File: .github/workflows/build_and_deploy.yml [hunk 5]
```diff
@@ -84,7 +84,7 @@ jobs:
     strategy:
       max-parallel: 1
       matrix:
-        environment: [qa, sandbox] ## add future nonprod environments here
+        environment: [qa, sandbox, production] ## add future nonprod environments here
```
Comment (RMcVelia) [inline .github/workflows/build_and_deploy.yml:87]: production would have a separate job i.e. deploy_prod, which would require deploy_nonprod

File: terraform/application/config/production_app_env.yml [hunk 6]
```diff
@@ -0,0 +1,10 @@
+---
+SIGN_IN_METHOD: dfe-sign-in
+GIAS_CSV_BASE_URL: https://ea-edubase-api-prod.azurewebsites.net/edubase/downloads/public
+PUBLISH_BASE_URL: https://api.publish-teacher-training-courses.service.gov.uk/
+HOSTING_ENV: production
+TEACHING_RECORD_BASE_URL: https://teacher-qualifications-api.education.gov.uk/
+TEACHING_RECORD_API_MINOR_VERSION: 20240101
+CLAIMS_DFE_SIGN_IN_CLIENT_ID: ittMentorTrackPay
+DFE_SIGN_IN_ISSUER_URL: https://oidc.signin.education.gov.uk
+PLACEMENTS_DFE_SIGN_IN_CLIENT_ID: ittMentorSchoolPlacement
```
Comment (RMcVelia) [inline terraform/application/config/production_app_env.yml:10]: need to add CLAIMS_HOSTS

File: terraform/domains/environment_domains/config/production.tfvars.json [hunk 7]
```diff
@@ -0,0 +1,16 @@
+{
+    "hosted_zone": {
+      "claim-funding-for-mentor-training.education.gov.uk": {
+        "front_door_name": "s189p01-ittmsdomains-fd",
+        "resource_group_name": "s189p01-ittmsdomains-rg",
+        "domains": [
+          "production"
```
[the remark this answers is not in the document]
Comment (temitope777) [reply terraform/domains/environment_domains/config/production.tfvars.json:7]: changed to apex

File: terraform/domains/environment_domains/config/production.tfvars.json [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+        ],
+        "cached_paths": [
+          "/assets/*"
+        ],
+        "environment_short": "pd",
+        "origin_hostname": "track-and-pay-sandbox.teacherservices.cloud"
```
[the remark this answers is not in the document]
Comment (temitope777) [reply terraform/domains/environment_domains/config/production.tfvars.json:13]: changed

File: global_config/production.sh (hunk 1 above)
Comment (temitope777) [reply global_config/production.sh]: changed to production

File: .github/workflows/build_and_deploy.yml (hunk 5 above)
Comment (temitope777) [reply .github/workflows/build_and_deploy.yml]: refactored

File: terraform/application/config/production_Terrafile [hunk 9]
```diff
@@ -0,0 +1,6 @@
+aks:
+    source:  "https://github.com/DFE-Digital/terraform-modules"
+    version: "stable"
+
```
Comment (Nitemaeric) [inline terraform/application/config/production_Terrafile:4]: Do you want to clean this file up before you merge?

File: Makefile [hunk 10]
```diff
@@ -42,6 +42,14 @@ staging: test-cluster
 		$(eval export TF_VAR_environment=${ENVIRONMENT})
 		$(eval include global_config/staging.sh)
 
+
```
Comment (Nitemaeric) [inline Makefile:45]: I think the convention is a single empty newline between commands.

File: Makefile [hunk 11]
```diff
@@ -42,6 +42,14 @@ staging: test-cluster
...
+.PHONY: production
+production: production-cluster
+		$(eval ENVIRONMENT=production)
+		$(eval export TF_VAR_environment=${ENVIRONMENT})
+		$(eval include global_config/production.sh)
+
+
```
Comment (Nitemaeric) [inline Makefile:52]: Here too.

File: Makefile [hunk 12]
```diff
@@ -166,3 +174,6 @@ get-cluster-credentials: set-azure-account
 bin/konduit.sh:
 	curl -s https://raw.githubusercontent.com/DFE-Digital/teacher-services-cloud/main/scripts/konduit.sh -o bin/konduit.sh \
 		&& chmod +x bin/konduit.sh
+
+
+
```
Comment (Nitemaeric) [inline Makefile:179]: These two lines should be removed.

**Review by Nitemaeric — APPROVED** [review]
Code looks good to me. Maybe a linter could assist with ensuring consistent code styling?

File: Makefile (hunk 10 above)
Comment (temitope777) [reply Makefile:45]: double space removed

File: Makefile (hunk 11 above)
Comment (temitope777) [reply Makefile:52]: removed

File: Makefile (hunk 12 above)
Comment (temitope777) [reply Makefile:179]: lines removed

File: terraform/application/variables.tf [hunk 13]
```diff
@@ -41,7 +41,41 @@ variable "enable_monitoring" {
...
+  default = null 
+}
+variable "postgres_flexible_server_sku" { 
+  default = "B_Standard_B1ms" 
+}
+variable "postgres_enable_high_availability" { 
+  default = false 
+}
+variable "azure_enable_backup_storage" { 
+  default = true
```
Comment (RMcVelia) [inline terraform/application/variables.tf:74]: default should be false, and set to true for prod env only for now

File: terraform/domains/environment_domains/config/production.tfvars.json [hunk 14]
```diff
@@ -0,0 +1,22 @@
...
+        "origin_hostname": "track-and-pay-production.teacherservices.cloud",
+        "redirect_rules": [
+          {
+            "from-domain": "www",
+            "to-domain": "claim-funding-for-mentor-training.education.gov.uk"
+          }
+        ]
+      }
+    }
+  }
```
Comment (RMcVelia) [inline terraform/domains/environment_domains/config/production.tfvars.json:22]: think you need to add redirect_rules to the main.tf for environment domains, as I don't see a redirect rule in fd
i.e.
redirect_rules      = try(each.value.redirect_rules, [])

Comment (temitope777) [reply terraform/domains/environment_domains/config/production.tfvars.json:22]: added

**Review by RMcVelia — APPROVED** [review]

**Comment by temitope777** [comment]
[ITTMS: Create production environment](https://trello.com/c/3jvfiAei/1667-ittms-create-production-environment)

**Pull request merged**